File: Backend/shipment/shipment/src/main/java/com/shipfast/shipment/entity/Shipment.java
package com.shipfast.shipment.entity;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Document(collection = "shipments")
public class Shipment {

    @Id
    private String id;

    private String trackingNumber;
    private String customerId;
    private String branchId;

    private String status;
    private String serviceType;
    private String paymentMethod;
    private String paymentStatus;

    private Double cost;

    private LocalDateTime createdAt;
    private LocalDateTime estimatedDelivery;
    private LocalDateTime updatedAt;
    private LocalDateTime paymentCollectedAt;

    private Address sender;
    private Address recipient;

    private PackageDetails packageDetails;

    @Builder.Default
    private List<TrackingEvent> history = new ArrayList<>();

    private String assignedAgentId;
    private Integer rating;
    private String ratingComment;
    private String proofOfDeliveryImage;
    private LocalDateTime deliveredAt;
    private String deliveredBy;
    private String deliveredByAgentId;
    private String runSheetId;
}

File: mern-backend/src/models/Shipment.js
import { createModel } from '../db/mongo.js';

const ShipmentMethods = {
  toDto: function () {
    return {
      id: this.id || this.shipmentId || this.trackingNumber || this._id?.toString() || '',
      shipmentId: this.id || this.shipmentId || this.trackingNumber || this._id?.toString() || '',
      trackingNumber: this.trackingNumber || this.id || '',
      trackingId: this.trackingNumber || this.id || '',
      customerId: this.customerId || '',
      branchId: this.branchId || '',
      status: this.status || 'Booked',
      serviceType: this.serviceType || 'Standard',
      paymentMethod: this.paymentMethod || 'ONLINE',
      paymentStatus: this.paymentStatus || 'SUCCESS',
      paymentCollectedAt: this.paymentCollectedAt || null,
      cost: Number(this.cost || 0),
      estimatedDelivery: this.estimatedDelivery || null,
      deliveredAt: this.deliveredAt || null,
      deliveredBy: this.deliveredBy || null,
      deliveredByAgentId: this.deliveredByAgentId || null,
      assignedAgentId: this.assignedAgentId || null,
      runSheetId: this.runSheetId || null,
      proofOfDeliveryImage: this.proofOfDeliveryImage || null,
      rating: this.rating != null ? Number(this.rating) : null,
      ratingComment: this.ratingComment || '',
      sender: this.sender || {},
      recipient: this.recipient || {},
      packageDetails: this.packageDetails || {},
      history: Array.isArray(this.history) ? this.history : [],
      createdAt: this.createdAt || new Date(),
      updatedAt: this.updatedAt || new Date()
    };
  }
};

const ShipmentDefaults = {
  customerId: '',
  branchId: '',
  status: 'Booked',
  serviceType: 'Standard',
  paymentMethod: 'ONLINE',
  paymentStatus: 'SUCCESS',
  paymentCollectedAt: null,
  cost: 0,
  estimatedDelivery: null,
  deliveredAt: null,
  deliveredBy: null,
  deliveredByAgentId: null,
  assignedAgentId: null,
  runSheetId: null,
  proofOfDeliveryImage: null,
  rating: null,
  ratingComment: '',
  sender: {},
  recipient: {},
  packageDetails: {},
  history: []
};

export const Shipment = createModel('shipments', ShipmentMethods, ShipmentDefaults);
